{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6f6b9257",
   "metadata": {},
   "source": [
    "# Computer Assignment #2 Game Report\n",
    "# AI course - Fall 1400\n",
    "<h2>By Kianoush Arshi<br>Std ID: 810198438</h2>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cdfb866",
   "metadata": {},
   "source": [
    "<h2>Part 1: Genes and Chromosomes</h2>\n",
    "In this algorithm we have 3 different classes of data:<br>\n",
    "<li>Genome, which is the smallest and mine is each cell of the board.<br>\n",
    "<li>Chromosme, which is a cluster of genes together and mine is a 9 in 9 board.<br>\n",
    "<li>Population, which is a list of chromosmes and my population size is 200.<br></li>\n",
    "\n",
    "<h2>Genome</h2>\n",
    "<p>Each cell is a genome, in the algorithms mutation function we swap 2 genes in the same chromosome and in the cross over we swap a group of genes between two chromosomes.<br></p>\n",
    "\n",
    "<h2>Chromosome</h2>\n",
    "<p>\n",
    "    Each chromosome is a possible solution to our problem.<br>\n",
    "    We can find out if a solution is found using a fitness function which is defined as follows:<br>\n",
    "    <li>Get the count of duplicate numbers in each column and divide them by (9*9)</li>\n",
    "    <li>Get the count of duplicate numbers in each 3by3 block and divide them by (9*9)</li>\n",
    "    <li>Check if there is any duplicates:\n",
    "        <li>If there isn't any: set fitness to 1 and return.\n",
    "        <li>If there is: return the multitude of block_sum and col_sum as fitness.\n",
    "    </li>    \n",
    "</p>\n",
    "\n",
    "<h2>Population</h2>\n",
    "<p>\n",
    "    We have a ppulation size of 200. Initial population is generated randomly with the only two constraints of not having duplicated any of the fixed values in a block, row or column and not inputted more than 9 times.<br>\n",
    "    Each generation, we take 20% of the top(fittesst) chromosomes and keep them as elites in the next generation. Using these elites, we create children by randomly crossing over two of them and possibly mutating the child.<br>\n",
    "    <h4>Cross Over:</h4>\n",
    "    <p>\n",
    "        A child is created using a cross over between two randomly selected chromosomes and then using two randomly generated cross over points, switched genes between these points in each row.\n",
    "        The aforementioned approach for cross over had some problems, only the elites had a chance to be a parent. This might cause the algorithm to get stuck in local minima. Thus a new approach is introduced:<br>\n",
    "        Have a competition between two randomly chosen chromosomes and give the weaker one a chance to be chosen reducing the chance for the population to go stale.<br>\n",
    "        Afterwards, the cross over is as follows:<br>\n",
    "    <li>Randomly choose two cross over point.</li>\n",
    "    <li>Apply cyclic cross over to rows from cross over point 1 to cross over point 2.</li>\n",
    "    </p>\n",
    "    <h4>Muation:</h4>\n",
    "    <p>\n",
    "        In general, optimization solving algorithms converge to a local minimum.  To get out of this local minimum in a Genetic Algorithm, you can use mutations. The higher the mutation rate is, the more 'space' will be searched and the higher the chance that the global minimum is found. But if the mutation rate is too high, the algorithm won't converge anymore.<br>\n",
    "        The cross overed child is then mutated with an adaptive mutation rate which chaanges each time we encounter a successful mutation and resets each time we encounter stale population.<br>\n",
    "        Significant progress toward the fitness function's optimum, generally, can only happen in a narrow band of the mutation step size sigma. That narrow band is called evolution window.<br>\n",
    "        There are three well-known methods to adapt the mutation step size Ïƒ in evolution strategies:<br>\n",
    "            <li>(1/5-th) Success Rule</li>\n",
    "            <li>Self-Adaptation (for example through log-normal mutations)</li>\n",
    "            <li>Cumulative Step Size Adaptation (CSA)</li>\n",
    "    The evolution strategy utilized in this project is the 1/5-th success rule.\n",
    "    </p>\n",
    "    <h4>Handling Stale Population:</h4>\n",
    "    <p>\n",
    "        To detect a stale population, we have a counter which counts each time we see that the two best chromosomes have the same fitness. If this counter reaches 100, we reset the algorithm with new initial population and the base mutation rate.\n",
    "    </p>\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a2b8038e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_file(filename : str) -> list() :\n",
    "    file = open(filename, 'r+')\n",
    "    init_board = []\n",
    "    for line in file.readlines():\n",
    "        init_board.append([int(i) for i in line.split()])\n",
    "        \n",
    "    return init_board\n",
    "\n",
    "def print_chromosome(chromosome):\n",
    "    if chromosome == None:\n",
    "        return\n",
    "    for i in range(len(chromosome)):\n",
    "        for j in range(len(chromosome[0])):\n",
    "            print(chromosome[i][j], end = \" \")\n",
    "        print()\n",
    "    print()    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2da80845",
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_row_duplicate(row, value, chromosome):\n",
    "    for column in range(0, 9):\n",
    "        if chromosome[row][column] == value:\n",
    "            return True\n",
    "    return False\n",
    "\n",
    "def is_column_duplicate(column, value, chromosome):\n",
    "    for row in range(0, 9):\n",
    "        if chromosome[row][column] == value:\n",
    "            return True\n",
    "    return False\n",
    "\n",
    "def is_block_duplicate(row, column, value, chromosome):\n",
    "    i = 3 * (int(row / 3))\n",
    "    j = 3 * (int(column / 3))\n",
    "\n",
    "    if ((chromosome[i][j] == value)\n",
    "        or (chromosome[i][j + 1] == value)\n",
    "        or (chromosome[i][j + 2] == value)\n",
    "        or (chromosome[i + 1][j] == value)\n",
    "        or (chromosome[i + 1][j + 1] == value)\n",
    "        or (chromosome[i + 1][j + 2] == value)\n",
    "        or (chromosome[i + 2][j] == value)\n",
    "        or (chromosome[i + 2][j + 1] == value)\n",
    "        or (chromosome[i + 2][j + 2] == value)):\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "def make_index(v):\n",
    "    if v <= 2:\n",
    "        return 0\n",
    "    elif v <= 5:\n",
    "        return 3\n",
    "    else:\n",
    "        return 6\n",
    "    \n",
    "def find_unused(parent_row, remaining):\n",
    "    for i in range(0, len(parent_row)):\n",
    "        if (parent_row[i] in remaining):\n",
    "            return i\n",
    "\n",
    "def find_value(parent_row, value):\n",
    "    for i in range(0, len(parent_row)):\n",
    "        if (parent_row[i] == value):\n",
    "            return i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8d2a6db7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compete(chromosome1, chromosome2):\n",
    "\n",
    "        # Find the fittest and the weakest.\n",
    "        if (chromosome1[0] > chromosome2[0]):\n",
    "            fittest = chromosome1[1]\n",
    "            weakest = chromosome2[1]\n",
    "        else:\n",
    "            fittest = chromosome2[1]\n",
    "            weakest = chromosome1[1]\n",
    "\n",
    "        selection_rate = 0.80\n",
    "        r = random.uniform(0, 1.1)\n",
    "        while (r > 1):  # Outside [0, 1] boundary. Choose another.\n",
    "            r = random.uniform(0, 1.1)\n",
    "        if (r < selection_rate):\n",
    "            return fittest\n",
    "        else:\n",
    "            return weakest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7b18add4",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "import random\n",
    "from copy import deepcopy\n",
    "from math import floor\n",
    "import numpy as np\n",
    "\n",
    "random.seed()\n",
    "\n",
    "class SudokuSolver():\n",
    "\n",
    "    def __init__(self, init_board):\n",
    "        self.fixed_values = init_board\n",
    "        self.solved = 0\n",
    "        self.pop_size = 200  # population size\n",
    "        self.elite_percentage = 10 # keep x% best of parents in the new population\n",
    "        self.Ng = 10000  # Number of generations.\n",
    "        self.m_count = 0  # Number of mutations.\n",
    "        self.ans = []\n",
    "        \n",
    "        # Cross over probability\n",
    "        self.pc = 0.7\n",
    "        # Mutation parameters.\n",
    "        self.phi = 0\n",
    "        self.sigma = 1\n",
    "        self.pm = 0.06\n",
    "\n",
    "        # Create an initial population.\n",
    "        self.chromosomes = self.generate_initial_population()\n",
    "        return\n",
    "    \n",
    "    def generate_initial_population(self):\n",
    "        self.stale = 0\n",
    "        init_pop = []\n",
    "        \n",
    "        # Determine the legal values that each square can take.\n",
    "        helper = [[[] for j in range(0, 9)] for i in range(0, 9)]\n",
    "        for row in range(0, 9):\n",
    "            for column in range(0, 9):\n",
    "                for value in range(1, 10):\n",
    "                    if ((self.fixed_values[row][column] == 0) and not \n",
    "                        (is_column_duplicate(column, value, self.fixed_values) or \n",
    "                         is_block_duplicate(row, column, value, self.fixed_values) or \n",
    "                         is_row_duplicate(row, value, self.fixed_values))):\n",
    "                        # Value is available.\n",
    "                        helper[row][column].append(value)\n",
    "                    elif self.fixed_values[row][column] != 0:\n",
    "                        # Given/known value from file.\n",
    "                        helper[row][column].append(self.fixed_values[row][column])\n",
    "                        break\n",
    "\n",
    "        # Seed a new population.\n",
    "        for p in range(0, self.pop_size):\n",
    "            chromosome = []\n",
    "            for i in range(0, 9):  # New row in candidate.\n",
    "                row = [0]*9\n",
    "\n",
    "                # Fill in the givens.\n",
    "                for j in range(0, 9):  # New column j value in row i.\n",
    "\n",
    "                    # If value is already given, don't change it.\n",
    "                    if self.fixed_values[i][j] != 0:\n",
    "                        row[j] = self.fixed_values[i][j]\n",
    "                    # Fill in the gaps using the helper board.\n",
    "                    elif self.fixed_values[i][j] == 0:\n",
    "                        row[j] = helper[i][j][random.randint(0, len(helper[i][j]) - 1)]\n",
    "\n",
    "                # If we don't have a valid board, then try again. max iteration 500,000\n",
    "                # There must be no duplicates in the row.\n",
    "                ii = 0\n",
    "                while len(list(set(row))) != 9:\n",
    "                    ii += 1\n",
    "                    if ii > 500000:\n",
    "                        return 0\n",
    "                    for j in range(0, 9):\n",
    "                        if self.fixed_values[i][j] == 0:\n",
    "                            row[j] = helper[i][j][random.randint(0, len(helper[i][j]) - 1)]\n",
    "\n",
    "                chromosome.append(row)\n",
    "            init_pop.append(chromosome)\n",
    "\n",
    "        print_chromosome(init_pop[0])\n",
    "        print(\"Initial fitness:\", self.get_fitness(init_pop[0]))\n",
    "        return init_pop\n",
    "    \n",
    "    def cross_over(self, chromosome1, chromosome2):\n",
    "        # Make a copy of the parent genes.\n",
    "        child1 = deepcopy(chromosome1)\n",
    "        child2 = deepcopy(chromosome2)\n",
    "\n",
    "        # Pick a crossover point. Crossover must have at least 1 row (and at most 8) rows.\n",
    "        crossover_point1 = random.randint(0, 8)\n",
    "        crossover_point2 = random.randint(1, 9)\n",
    "        while (crossover_point1 == crossover_point2):\n",
    "            crossover_point1 = random.randint(0, 8)\n",
    "            crossover_point2 = random.randint(1, 9)\n",
    "\n",
    "        if (crossover_point1 > crossover_point2):\n",
    "            temp = crossover_point1\n",
    "            crossover_point1 = crossover_point2\n",
    "            crossover_point2 = temp\n",
    "            \n",
    "        for i in range(crossover_point1, crossover_point2):\n",
    "            child1[i], child2[i] = self.crossover_rows(child1[i], child2[i])\n",
    "\n",
    "        return child1, child2\n",
    "                    \n",
    "    def crossover_rows(self, row1, row2):\n",
    "        child_row1 = [0]*9\n",
    "        child_row2 = [0]*9\n",
    "\n",
    "        remaining = [i for i in range(1, 10)]\n",
    "        cycle = 0\n",
    "\n",
    "        while ((0 in child_row1) and (0 in child_row2)):\n",
    "            if (cycle % 2 == 0):  # Even cycles.\n",
    "                # Assign next unused value.\n",
    "                index = find_unused(row1, remaining)\n",
    "                start = row1[index]\n",
    "                remaining.remove(row1[index])\n",
    "                child_row1[index] = row1[index]\n",
    "                child_row2[index] = row2[index]\n",
    "                next = row2[index]\n",
    "\n",
    "                while (next != start):  # While cycle not done...\n",
    "                    index = find_value(row1, next)\n",
    "                    child_row1[index] = row1[index]\n",
    "                    remaining.remove(row1[index])\n",
    "                    child_row2[index] = row2[index]\n",
    "                    next = row2[index]\n",
    "\n",
    "                cycle += 1\n",
    "\n",
    "            else:  # Odd cycle - flip values.\n",
    "                index = find_unused(row1, remaining)\n",
    "                start = row1[index]\n",
    "                remaining.remove(row1[index])\n",
    "                child_row1[index] = row2[index]\n",
    "                child_row2[index] = row1[index]\n",
    "                next = row2[index]\n",
    "\n",
    "                while (next != start):  # While cycle not done...\n",
    "                    index = find_value(row1, next)\n",
    "                    child_row1[index] = row2[index]\n",
    "                    remaining.remove(row1[index])\n",
    "                    child_row2[index] = row1[index]\n",
    "                    next = row2[index]\n",
    "\n",
    "                cycle += 1\n",
    "\n",
    "        return child_row1, child_row2\n",
    "\n",
    "    def mutate(self, chromosome):\n",
    "        prob = random.random()\n",
    "        if (prob < 0.06):  # Mutate.\n",
    "            row1, col1 = (random.randint(0, 8), random.randint(0, 8))\n",
    "            row2, col2 = (random.randint(0, 8), random.randint(0, 8))\n",
    "\n",
    "            while (row1, col1) == (row2, col2):\n",
    "                row1, col1 = (random.randint(0, 8), random.randint(0, 8))\n",
    "                row2, col2 = (random.randint(0, 8), random.randint(0, 8))\n",
    "            row2 = row1\n",
    "\n",
    "            # Check if the two places are not fixed\n",
    "            if self.fixed_values[row1][col1] == 0 and self.fixed_values[row2][col2] == 0:\n",
    "                temp_chromosome = deepcopy(chromosome)\n",
    "                temp = temp_chromosome[row2][col2]\n",
    "                temp_chromosome[row2][col2] = temp_chromosome[row1][col1]\n",
    "                temp_chromosome[row1][col1] = temp\n",
    "                # and we are not lowering fitness.\n",
    "                if self.get_fitness(temp_chromosome) > self.get_fitness(chromosome):\n",
    "                    self.m_count += 1\n",
    "                    self.phi += 1\n",
    "                    chromosome = deepcopy(temp_chromosome)\n",
    "\n",
    "        return chromosome\n",
    "        \n",
    "        \n",
    "    def get_fitness(self, chromosome):\n",
    "        row_count = [0]*9\n",
    "        column_count = [0]*9\n",
    "        block_count = [0]*9\n",
    "        row_sum = 0\n",
    "        column_sum = 0\n",
    "        block_sum = 0\n",
    "        \n",
    "        # For each column....\n",
    "        for j in range(0, 9):\n",
    "            for i in range(0, 9):\n",
    "                column_count[chromosome[i][j] - 1] += 1\n",
    "                \n",
    "            for i in range(0, 9):\n",
    "                row_count[chromosome[j][i] - 1] += 1\n",
    "\n",
    "            for k in range(9):\n",
    "                if column_count[k] == 1:\n",
    "                    column_sum += 1\n",
    "                if row_count[k] == 1:\n",
    "                    row_sum += 1\n",
    "            column_count = [0]*9\n",
    "            row_count = [0]*9\n",
    "\n",
    "        # For each block...\n",
    "        for i in range(0, 9, 3):\n",
    "            for j in range(0, 9, 3):\n",
    "                block_count[chromosome[i][j] - 1] += 1\n",
    "                block_count[chromosome[i][j + 1] - 1] += 1\n",
    "                block_count[chromosome[i][j + 2] - 1] += 1\n",
    "\n",
    "                block_count[chromosome[i + 1][j] - 1] += 1\n",
    "                block_count[chromosome[i + 1][j + 1] - 1] += 1\n",
    "                block_count[chromosome[i + 1][j + 2] - 1] += 1\n",
    "\n",
    "                block_count[chromosome[i + 2][j] - 1] += 1\n",
    "                block_count[chromosome[i + 2][j + 1] - 1] += 1\n",
    "                block_count[chromosome[i + 2][j + 2] - 1] += 1\n",
    "\n",
    "                for k in range(9):\n",
    "                    if block_count[k] == 1:\n",
    "                        block_sum += 1\n",
    "                block_count = [0]*9\n",
    "\n",
    "        return column_sum + block_sum + row_sum\n",
    "\n",
    "    \n",
    "    def generate_new_population(self, generation):\n",
    "        sorted_chromosomes = [(self.get_fitness(chromosome), chromosome) for chromosome in self.chromosomes]\n",
    "        sorted_chromosomes.sort(reverse=True)\n",
    "        \n",
    "        # Check for stale population.\n",
    "        if (sorted_chromosomes[0] != sorted_chromosomes[1]):\n",
    "            self.stale = 0\n",
    "        else:\n",
    "            self.stale += 1\n",
    "\n",
    "        # Re-seed the population if 100 generations have passed\n",
    "        # with the fittest two candidates always having the same fitness.\n",
    "        if (self.stale >= 100):\n",
    "            print(\"The population has gone stale. Re-seeding...\")\n",
    "            self.sigma = 1\n",
    "            self.phi = 0\n",
    "            self.pm = 0.06\n",
    "            return self.generate_initial_population()\n",
    "\n",
    "        elites = sorted_chromosomes[:floor((self.elite_percentage / 100) * (self.pop_size))]\n",
    "        print(\"Generation {}. Best fitness: {}. Second best: {}\".format(generation, elites[0][0], elites[1][0]))\n",
    "        if (sorted_chromosomes[0][0] == 243): #no intersections in the best chromosome\n",
    "            self.solved = 1\n",
    "            self.ans = sorted_chromosomes[0][1]\n",
    "            return sorted_chromosomes[0]\n",
    "\n",
    "        new_pop = [chromosome[1] for chromosome in elites] # add elites\n",
    "\n",
    "        # Create the rest of the candidates.\n",
    "        while (len(new_pop) < self.pop_size):\n",
    "            # Select parents from population via a tournament.\n",
    "            potential_parents = random.sample(sorted_chromosomes, 4)\n",
    "            parent1 = compete(potential_parents[0], potential_parents[1])\n",
    "            parent2 = compete(potential_parents[2], potential_parents[3])\n",
    "\n",
    "            child1, child2 = self.cross_over(parent1, parent2)\n",
    "\n",
    "            mutated_child1 = self.mutate(child1)\n",
    "            mutated_child2 = self.mutate(child2)\n",
    "            new_pop.append(mutated_child1)\n",
    "            new_pop.append(mutated_child2)\n",
    "\n",
    "        return new_pop\n",
    "\n",
    "    def solve(self):        \n",
    "        for generation in range(0, self.Ng):\n",
    "            # Create the next population.\n",
    "            self.chromosomes = self.generate_new_population(generation)\n",
    "            # Calculate new adaptive mutation rate (based on Rechenberg's 1/5 success rule).\n",
    "            # This is to stop too much mutation as the fitness progresses towards unity.\n",
    "            if (self.m_count == 0):\n",
    "                self.phi = 0  # Avoid divide by zero.\n",
    "            else:\n",
    "                self.phi = self.phi / self.m_count\n",
    "\n",
    "            if (self.phi > 0.2):\n",
    "                self.sigma = self.sigma / 0.998\n",
    "            elif (self.phi < 0.2):\n",
    "                self.sigma = self.sigma * 0.998\n",
    "                \n",
    "            self.pm = abs(np.random.normal(loc=0.0, scale=self.sigma, size=None))\n",
    "                \n",
    "            if self.solved:\n",
    "                return self.ans\n",
    "\n",
    "        print(\"No solution found.\")\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6279a558",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'Nd' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m~\\AppData\\Local\\Temp/ipykernel_1876/3753591261.py\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      2\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 4\u001b[1;33m \u001b[0msudoku_solver\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mSudokuSolver\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mread_file\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"Test1.txt\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      5\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      6\u001b[0m \u001b[0mstart\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Local\\Temp/ipykernel_1876/1535071274.py\u001b[0m in \u001b[0;36m__init__\u001b[1;34m(self, init_board)\u001b[0m\n\u001b[0;32m     25\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     26\u001b[0m         \u001b[1;31m# Create an initial population.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 27\u001b[1;33m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mchromosomes\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mgenerate_initial_population\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     28\u001b[0m         \u001b[1;32mreturn\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     29\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Local\\Temp/ipykernel_1876/1535071274.py\u001b[0m in \u001b[0;36mgenerate_initial_population\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m     67\u001b[0m                 \u001b[1;31m# There must be no duplicates in the row.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     68\u001b[0m                 \u001b[0mii\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;36m0\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 69\u001b[1;33m                 \u001b[1;32mwhile\u001b[0m \u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mlist\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mset\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mrow\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m!=\u001b[0m \u001b[0mNd\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     70\u001b[0m                     \u001b[0mii\u001b[0m \u001b[1;33m+=\u001b[0m \u001b[1;36m1\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     71\u001b[0m                     \u001b[1;32mif\u001b[0m \u001b[0mii\u001b[0m \u001b[1;33m>\u001b[0m \u001b[1;36m500000\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'Nd' is not defined"
     ]
    }
   ],
   "source": [
    "from time import time\n",
    "\n",
    "\n",
    "sudoku_solver = SudokuSolver(read_file(\"Test1.txt\"))\n",
    "\n",
    "start = time()\n",
    "\n",
    "print_chromosome(sudoku_solver.solve())\n",
    "\n",
    "end = time()\n",
    "\n",
    "print(\"Time taken:\", (end-start))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e7525db",
   "metadata": {},
   "outputs": [],
   "source": [
    "sudoku_solver = SudokuSolver(read_file(\"Test2.txt\"))\n",
    "\n",
    "start = time()\n",
    "\n",
    "print_chromosome(sudoku_solver.solve())\n",
    "\n",
    "end = time()\n",
    "\n",
    "print(\"Time taken:\", (end-start))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
